src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/Design/DesigntimeLicenseContextSerializer.cs

```diff
@@ -25,25 +24,97 @@ private DesigntimeLicenseContextSerializer()
         /// </summary>
         public static void Serialize(Stream o, string cryptoKey, DesigntimeLicenseContext context)
         {
-            IFormatter formatter = new BinaryFormatter();
+            if (LocalAppContextSwitches.BinaryFormatterEnabled)
+            {
+                IFormatter formatter = new BinaryFormatter();
 #pragma warning disable SYSLIB0011 // Issue https://github.com/dotnet/runtime/issues/39293 tracks finding an alternative to BinaryFormatter
-            formatter.Serialize(o, new object[] { cryptoKey, context._savedLicenseKeys });
+                formatter.Serialize(o, new object[] { cryptoKey, context._savedLicenseKeys });
 #pragma warning restore SYSLIB0011
+            }
+            else
+            {
+                using (BinaryWriter writer = new BinaryWriter(o, encoding: Text.Encoding.UTF8, leaveOpen: true))
+                {
+                    writer.Write(-1); // flag to identify BinaryWriter
+                    writer.Write(cryptoKey);
+                    writer.Write(context._savedLicenseKeys.Count);
+                    foreach (DictionaryEntry keyAndValue in context._savedLicenseKeys)
+                    {
+                        writer.Write(keyAndValue.Key.ToString());
+                        writer.Write(keyAndValue.Value.ToString());
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// During deserialization, the stream passed in may be binary formatted or may have used binary writer. This is quick test to discern between them.
+        /// </summary>
+        internal static bool StreamIsBinaryFormatted(Stream stream)
+        {
+            if (!stream.CanSeek)
+            {
+                throw new Exception("Expected a seek-able stream");
+            }
+            bool ret = true;
+
+            // For binary formatter, the first byte is the SerializationHeaderRecord and has a value 0
+            int firstByte = stream.ReadByte();
+            if (firstByte != 0)
+            {
+                ret = false;
```

Comment:
Should you assert the magic number you're using for the non BF case? Looks like -1. And throw otherwise. (You could pick something more exotic than -1 possibly.)

That way you could give a better error than later when you try to read garbage.

Reply:
Yup, good idea. Added the assert. `-1` is good enough though, because BinaryFormatter can never start with `-1` as it's first byte.